File: OTCDerivativesCalculatorModule/Project_Cpp/lib_static/FpmlSerialized/GenClass/fpml-eq-shared-5-4/Compounding.cpp
// Compounding.cpp 
#include "Compounding.hpp"
#ifdef ConsolePrint
    #include <iostream>
#endif 

namespace FpmlSerialized {

Compounding::Compounding(TiXmlNode* xmlNode)
: ISerialized(xmlNode)
{
    #ifdef ConsolePrint
        std::string initialtap_ = FileManager::instance().tap_;
        FileManager::instance().tap_.append("   ");
    #endif 
   //compoundingMethodNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* compoundingMethodNode = xmlNode->FirstChildElement("compoundingMethod");

   if(compoundingMethodNode){compoundingMethodIsNull_ = false;}
   else{compoundingMethodIsNull_ = true;}

   #ifdef ConsolePrint
      FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- compoundingMethodNode , address : " << compoundingMethodNode << std::endl;
   #endif
   if(compoundingMethodNode)
   {
      if(compoundingMethodNode->Attribute("href") || compoundingMethodNode->Attribute("id"))
      {
          if(compoundingMethodNode->Attribute("id"))
          {
             compoundingMethodIDRef_ = compoundingMethodNode->Attribute("id");
             compoundingMethod_ = boost::shared_ptr<CompoundingMethodEnum>(new CompoundingMethodEnum(compoundingMethodNode));
             compoundingMethod_->setID(compoundingMethodIDRef_);
             IDManager::instance().setID(compoundingMethodIDRef_,compoundingMethod_);
          }
          else if(compoundingMethodNode->Attribute("href")) { compoundingMethodIDRef_ = compoundingMethodNode->Attribute("href");}
          else { QL_FAIL("id or href error"); }
      }
      else { compoundingMethod_ = boost::shared_ptr<CompoundingMethodEnum>(new CompoundingMethodEnum(compoundingMethodNode));}
   }

   //compoundingRateNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* compoundingRateNode = xmlNode->FirstChildElement("compoundingRate");

   if(compoundingRateNode){compoundingRateIsNull_ = false;}
   else{compoundingRateIsNull_ = true;}

   #ifdef ConsolePrint
      FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- compoundingRateNode , address : " << compoundingRateNode << std::endl;
   #endif
   if(compoundingRateNode)
   {
      if(compoundingRateNode->Attribute("href") || compoundingRateNode->Attribute("id"))
      {
          if(compoundingRateNode->Attribute("id"))
          {
             compoundingRateIDRef_ = compoundingRateNode->Attribute("id");
             compoundingRate_ = boost::shared_ptr<CompoundingRate>(new CompoundingRate(compoundingRateNode));
             compoundingRate_->setID(compoundingRateIDRef_);
             IDManager::instance().setID(compoundingRateIDRef_,compoundingRate_);
          }
          else if(compoundingRateNode->Attribute("href")) { compoundingRateIDRef_ = compoundingRateNode->Attribute("href");}
          else { QL_FAIL("id or href error"); }
      }
      else { compoundingRate_ = boost::shared_ptr<CompoundingRate>(new CompoundingRate(compoundingRateNode));}
   }

   //compoundingSpreadNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* compoundingSpreadNode = xmlNode->FirstChildElement("compoundingSpread");

   if(compoundingSpreadNode){compoundingSpreadIsNull_ = false;}
   else{compoundingSpreadIsNull_ = true;}

   #ifdef ConsolePrint
      FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- compoundingSpreadNode , address : " << compoundingSpreadNode << std::endl;
   #endif
   if(compoundingSpreadNode)
   {
      if(compoundingSpreadNode->Attribute("href") || compoundingSpreadNode->Attribute("id"))
      {
          if(compoundingSpreadNode->Attribute("id"))
          {
             compoundingSpreadIDRef_ = compoundingSpreadNode->Attribute("id");
             compoundingSpread_ = boost::shared_ptr<XsdTypeDecimal>(new XsdTypeDecimal(compoundingSpreadNode));
             compoundingSpread_->setID(compoundingSpreadIDRef_);
             IDManager::instance().setID(compoundingSpreadIDRef_,compoundingSpread_);
          }
          else if(compoundingSpreadNode->Attribute("href")) { compoundingSpreadIDRef_ = compoundingSpreadNode->Attribute("href");}
          else { QL_FAIL("id or href error"); }
      }
      else { compoundingSpread_ = boost::shared_ptr<XsdTypeDecimal>(new XsdTypeDecimal(compoundingSpreadNode));}
   }

   //compoundingDatesNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* compoundingDatesNode = xmlNode->FirstChildElement("compoundingDates");

   if(compoundingDatesNode){compoundingDatesIsNull_ = false;}
   else{compoundingDatesIsNull_ = true;}

   #ifdef ConsolePrint
      FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- compoundingDatesNode , address : " << compoundingDatesNode << std::endl;
   #endif
   if(compoundingDatesNode)
   {
      if(compoundingDatesNode->Attribute("href") || compoundingDatesNode->Attribute("id"))
      {
          if(compoundingDatesNode->Attribute("id"))
          {
             compoundingDatesIDRef_ = compoundingDatesNode->Attribute("id");
             compoundingDates_ = boost::shared_ptr<AdjustableRelativeOrPeriodicDates2>(new AdjustableRelativeOrPeriodicDates2(compoundingDatesNode));
             compoundingDates_->setID(compoundingDatesIDRef_);
             IDManager::instance().setID(compoundingDatesIDRef_,compoundingDates_);
          }
          else if(compoundingDatesNode->Attribute("href")) { compoundingDatesIDRef_ = compoundingDatesNode->Attribute("href");}
          else { QL_FAIL("id or href error"); }
      }
      else { compoundingDates_ = boost::shared_ptr<AdjustableRelativeOrPeriodicDates2>(new AdjustableRelativeOrPeriodicDates2(compoundingDatesNode));}
   }

    #ifdef ConsolePrint
        FileManager::instance().tap_ = initialtap_;
    #endif 
}
boost::shared_ptr<CompoundingMethodEnum> Compounding::getCompoundingMethod()
{
   if(!this->compoundingMethodIsNull_){
        if(compoundingMethodIDRef_ != ""){
             return boost::shared_static_cast<CompoundingMethodEnum>(IDManager::instance().getID(compoundingMethodIDRef_));
        }else{
             return this->compoundingMethod_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return boost::shared_ptr<CompoundingMethodEnum>();
   }
}
boost::shared_ptr<CompoundingRate> Compounding::getCompoundingRate()
{
   if(!this->compoundingRateIsNull_){
        if(compoundingRateIDRef_ != ""){
             return boost::shared_static_cast<CompoundingRate>(IDManager::instance().getID(compoundingRateIDRef_));
        }else{
             return this->compoundingRate_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return boost::shared_ptr<CompoundingRate>();
   }
}
boost::shared_ptr<XsdTypeDecimal> Compounding::getCompoundingSpread()
{
   if(!this->compoundingSpreadIsNull_){
        if(compoundingSpreadIDRef_ != ""){
             return boost::shared_static_cast<XsdTypeDecimal>(IDManager::instance().getID(compoundingSpreadIDRef_));
        }else{
             return this->compoundingSpread_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return boost::shared_ptr<XsdTypeDecimal>();
   }
}
boost::shared_ptr<AdjustableRelativeOrPeriodicDates2> Compounding::getCompoundingDates()
{
   if(!this->compoundingDatesIsNull_){
        if(compoundingDatesIDRef_ != ""){
             return boost::shared_static_cast<AdjustableRelativeOrPeriodicDates2>(IDManager::instance().getID(compoundingDatesIDRef_));
        }else{
             return this->compoundingDates_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return boost::shared_ptr<AdjustableRelativeOrPeriodicDates2>();
   }
}
}

File: OTCDerivativesCalculatorModule/Project_CSharp/XmlBuilder/XmlSerializedClass/fpml-eq-shared-5-4/Compounding.cs
using System;
using System.Text;
using System.Collections.Generic;
using System.Xml;

namespace FpmlSerializedCSharp
{
    public class Compounding : ISerialized
    {
        public Compounding(XmlNode xmlNode)
        {
            XmlNode compoundingMethodNode = xmlNode.SelectSingleNode("compoundingMethod");
            
            if (compoundingMethodNode != null)
            {
                if (compoundingMethodNode.Attributes["href"] != null || compoundingMethodNode.Attributes["id"] != null) 
                {
                    if (compoundingMethodNode.Attributes["id"] != null) 
                    {
                        compoundingMethodIDRef_ = compoundingMethodNode.Attributes["id"].Value;
                        CompoundingMethodEnum ob = new CompoundingMethodEnum(compoundingMethodNode);
                        IDManager.SetID(compoundingMethodIDRef_, ob);
                    }
                    else if (compoundingMethodNode.Attributes["href"] != null)
                    {
                        compoundingMethodIDRef_ = compoundingMethodNode.Attributes["href"].Value;
                    }
                    else
                    {
                        compoundingMethod_ = new CompoundingMethodEnum(compoundingMethodNode);
                    }
                }
                else
                {
                    compoundingMethod_ = new CompoundingMethodEnum(compoundingMethodNode);
                }
            }
            
        
            XmlNode compoundingRateNode = xmlNode.SelectSingleNode("compoundingRate");
            
            if (compoundingRateNode != null)
            {
                if (compoundingRateNode.Attributes["href"] != null || compoundingRateNode.Attributes["id"] != null) 
                {
                    if (compoundingRateNode.Attributes["id"] != null) 
                    {
                        compoundingRateIDRef_ = compoundingRateNode.Attributes["id"].Value;
                        CompoundingRate ob = new CompoundingRate(compoundingRateNode);
                        IDManager.SetID(compoundingRateIDRef_, ob);
                    }
                    else if (compoundingRateNode.Attributes["href"] != null)
                    {
                        compoundingRateIDRef_ = compoundingRateNode.Attributes["href"].Value;
                    }
                    else
                    {
                        compoundingRate_ = new CompoundingRate(compoundingRateNode);
                    }
                }
                else
                {
                    compoundingRate_ = new CompoundingRate(compoundingRateNode);
                }
            }
            
        
            XmlNode compoundingSpreadNode = xmlNode.SelectSingleNode("compoundingSpread");
            
            if (compoundingSpreadNode != null)
            {
                if (compoundingSpreadNode.Attributes["href"] != null || compoundingSpreadNode.Attributes["id"] != null) 
                {
                    if (compoundingSpreadNode.Attributes["id"] != null) 
                    {
                        compoundingSpreadIDRef_ = compoundingSpreadNode.Attributes["id"].Value;
                        XsdTypeDecimal ob = new XsdTypeDecimal(compoundingSpreadNode);
                        IDManager.SetID(compoundingSpreadIDRef_, ob);
                    }
                    else if (compoundingSpreadNode.Attributes["href"] != null)
                    {
                        compoundingSpreadIDRef_ = compoundingSpreadNode.Attributes["href"].Value;
                    }
                    else
                    {
                        compoundingSpread_ = new XsdTypeDecimal(compoundingSpreadNode);
                    }
                }
                else
                {
                    compoundingSpread_ = new XsdTypeDecimal(compoundingSpreadNode);
                }
            }
            
        
            XmlNode compoundingDatesNode = xmlNode.SelectSingleNode("compoundingDates");
            
            if (compoundingDatesNode != null)
            {
                if (compoundingDatesNode.Attributes["href"] != null || compoundingDatesNode.Attributes["id"] != null) 
                {
                    if (compoundingDatesNode.Attributes["id"] != null) 
                    {
                        compoundingDatesIDRef_ = compoundingDatesNode.Attributes["id"].Value;
                        AdjustableRelativeOrPeriodicDates2 ob = new AdjustableRelativeOrPeriodicDates2(compoundingDatesNode);
                        IDManager.SetID(compoundingDatesIDRef_, ob);
                    }
                    else if (compoundingDatesNode.Attributes["href"] != null)
                    {
                        compoundingDatesIDRef_ = compoundingDatesNode.Attributes["href"].Value;
                    }
                    else
                    {
                        compoundingDates_ = new AdjustableRelativeOrPeriodicDates2(compoundingDatesNode);
                    }
                }
                else
                {
                    compoundingDates_ = new AdjustableRelativeOrPeriodicDates2(compoundingDatesNode);
                }
            }
            
        
        }
        
    
        #region CompoundingMethod_
        private CompoundingMethodEnum compoundingMethod_;
        public CompoundingMethodEnum CompoundingMethod_
        {
            get
            {
                if (this.compoundingMethod_ != null)
                {
                    return this.compoundingMethod_; 
                }
                else if (this.compoundingMethodIDRef_ != null)
                {
                    compoundingMethod_ = IDManager.getID(compoundingMethodIDRef_) as CompoundingMethodEnum;
                    return this.compoundingMethod_; 
                }
                else
                {
                      return this.compoundingMethod_; 
                }
            }
            set
            {
                if (this.compoundingMethod_ != value)
                {
                    this.compoundingMethod_ = value;
                }
            }
        }
        #endregion
        
        public string compoundingMethodIDRef_ { get; set; }
        #region CompoundingRate_
        private CompoundingRate compoundingRate_;
        public CompoundingRate CompoundingRate_
        {
            get
            {
                if (this.compoundingRate_ != null)
                {
                    return this.compoundingRate_; 
                }
                else if (this.compoundingRateIDRef_ != null)
                {
                    compoundingRate_ = IDManager.getID(compoundingRateIDRef_) as CompoundingRate;
                    return this.compoundingRate_; 
                }
                else
                {
                      return this.compoundingRate_; 
                }
            }
            set
            {
                if (this.compoundingRate_ != value)
                {
                    this.compoundingRate_ = value;
                }
            }
        }
        #endregion
        
        public string compoundingRateIDRef_ { get; set; }
        #region CompoundingSpread_
        private XsdTypeDecimal compoundingSpread_;
        public XsdTypeDecimal CompoundingSpread_
        {
            get
            {
                if (this.compoundingSpread_ != null)
                {
                    return this.compoundingSpread_; 
                }
                else if (this.compoundingSpreadIDRef_ != null)
                {
                    compoundingSpread_ = IDManager.getID(compoundingSpreadIDRef_) as XsdTypeDecimal;
                    return this.compoundingSpread_; 
                }
                else
                {
                      return this.compoundingSpread_; 
                }
            }
            set
            {
                if (this.compoundingSpread_ != value)
                {
                    this.compoundingSpread_ = value;
                }
            }
        }
        #endregion
        
        public string compoundingSpreadIDRef_ { get; set; }
        #region CompoundingDates_
        private AdjustableRelativeOrPeriodicDates2 compoundingDates_;
        public AdjustableRelativeOrPeriodicDates2 CompoundingDates_
        {
            get
            {
                if (this.compoundingDates_ != null)
                {
                    return this.compoundingDates_; 
                }
                else if (this.compoundingDatesIDRef_ != null)
                {
                    compoundingDates_ = IDManager.getID(compoundingDatesIDRef_) as AdjustableRelativeOrPeriodicDates2;
                    return this.compoundingDates_; 
                }
                else
                {
                      return this.compoundingDates_; 
                }
            }
            set
            {
                if (this.compoundingDates_ != value)
                {
                    this.compoundingDates_ = value;
                }
            }
        }
        #endregion
        
        public string compoundingDatesIDRef_ { get; set; }
        
    
        
    
    }
    
}

